hail_search/hail_search_query.py

```diff
@@ -0,0 +1,322 @@
+        annotation_fields.update({
+            enum_config.get('response_key', k): format_enum(k, enum_config)
+            for k, enum_config in self.ENUM_ANNOTATION_FIELDS.items()
+        })
+
+        if self._genome_version == GENOME_VERSION_GRCh38_DISPLAY:
+            annotation_fields.update(self.LIFTOVER_ANNOTATION_FIELDS)
+        return annotation_fields
+
+    def population_expression(self, r, population):
+        pop_config = self._format_population_config(self.POPULATIONS[population])
+        pop_field = self.POPULATION_FIELDS.get(population, population)
+        return hl.struct(**{
+            response_key: hl.or_else(r[pop_field][field], '' if response_key == 'id' else 0)
+            for response_key, field in pop_config.items() if field is not None
+        })
+
+    def _format_transcript_args(self):
+        return {
+            'format_value': lambda value: value.rename({k: _to_camel_case(k) for k in value.keys()}),
+        }
+
+    @staticmethod
+    def _enum_field(value, enum, ht_globals=None, annotate=None, format_value=None, drop_fields=None, **kwargs):
+        annotations = {}
+        drop = [] + (drop_fields or [])
+        value_keys = value.keys()
+        for field, field_enum in enum.items():
+            is_array = f'{field}_ids' in value_keys
+            value_field = f"{field}_id{'s' if is_array else ''}"
+            drop.append(value_field)
+
+            enum_array = hl.array(field_enum)
+            if is_array:
+                annotations[f'{field}s'] = value[value_field].map(lambda v: enum_array[v])
+            else:
+                annotations[field] = enum_array[value[value_field]]
+
+        value = value.annotate(**annotations)
+        if annotate:
+            annotations = annotate(value, enum, ht_globals)
+            value = value.annotate(**annotations)
+        value = value.drop(*drop)
+
+        if format_value:
+            value = format_value(value)
+
+        return value
+
+    def __init__(self, data_type, sample_data, genome_version, sort=XPOS_SORT_KEY, num_results=100, **kwargs):
+        self._genome_version = genome_version
+        self._sort = sort
+        self._num_results = num_results
+
+        self._load_filtered_table(data_type, sample_data, **kwargs)
+
+    def _load_filtered_table(self, data_type, sample_data, **kwargs):
+        self._ht = self.import_filtered_table(data_type, sample_data, **kwargs)
+
+    def import_filtered_table(self, data_type, sample_data, **kwargs):
+        tables_path = f'{DATASETS_DIR}/{self._genome_version}/{data_type}'
+
+        family_samples = defaultdict(list)
+        project_samples = defaultdict(list)
+        for s in sample_data:
+            family_samples[s['family_guid']].append(s)
+            project_samples[s['project_guid']].append(s)
+
+        logger.info(f'Loading {data_type} data for {len(family_samples)} families in {len(project_samples)} projects')
```

Comment:
The `project_samples` is solely for stats. A per-project counter should be enough.

Reply:
I do not understand. How is creating a tracker better than logging as we go so you can check the logs and see what is happeneing?